File: Lenguajes_Programacion/Java/6_funciones.java
/* FUNCIONES */

// Dentro de clas clases crean las funciones.
public class Main{
    public static void main(String[] args) {
        // Aqui se implementa la funcion sumar()
        Integer resultado = sumar(3, 4);
        System.out.println(resultado);
    }
    // Aqui se crea la funcion de sumar()
    static Integer sumar(Integer primero, Integer segundo) {
        return primero + segundo;
    }
}


// Ejemplo implementando funcion multiplicar()
public class Main{
    public static void main(String[] args) {
        int resultado = multiplicar(3, 4);
        System.out.println(resultado);
    }
    public static int multiplicar(int primero, int segundo) {
        return primero * segundo;
    }
}

// Ejemplo con una funcion de tipo "void" que no devulve ningun valor
public class Main{
    public static void main(String[] args) {
        String[] animales = {"perro", "gato", "tigre"};
        imprimirPrimerElemento(animales);
    }
    // void no devuleve ningun valor
    public static void imprimirPrimerElemento(String[] lista) {
        System.out.println(lista[0]);
    }
}

// Ejemplo con la funcion quicksort()

public class Main{
    public static void main(String[] args) {
        ArrayList<Integer> numeros = new ArrayList<Integer>();
        numeros.add(23);
        numeros.add(45);
        numeros.add(16);
        numeros.add(37);
        numeros.add(3);
        numeros.add(99);
        numeros.add(22);

        ArrayList<Integer> listaOrdenada = quicksort(numeros);
        System.out.println(listaOrdenada);
    }

    static ArrayList<Integer> quicksort(ArrayList<Integer> lista) {
        if (lista.size() <= 1) {
            return lista;
        }

        Integer pivote = lista.get(0);
        ArrayList<Integer> izquierda = new ArrayList<Integer>();
        ArrayList<Integer> derecha = new ArrayList<Integer>();
        for (int i = 1; i < lista.size(); i++) {
            if (lista.get(i) < pivote) {
                izquierda.add(lista.get(i));
            } else {
                derecha.add(lista.get(i));
            }
        }

        ArrayList<Integer> primero = quicksort(izquierda);
        ArrayList<Integer> medio = new ArrayList<Integer>();
        medio.add(pivote);
        ArrayList<Integer> segundo = quicksort(derecha);
        
        primero.addAll(medio);
        primero.addAll(segundo);
        return primero;
    };
}

File: Lenguajes_Programacion/3_C#/1_Sintaxis_C#/6_funciones.cs
/* Funciones */

// - Ejemplo 1:
int sumar(int primero, int segundo)
{
    return primero + segundo;
}

int resultado = sumar(1, 4);
Console.WriteLine(resultado);


// - Ejemplo 2:
numero1 = 23;

// No retorna ningun valor, ejecuta lo que se ha ingresado.
void multiplicar(int primero, int segundo)
{
    Console.WriteLine(primero + segundo);
}

multiplicar(numero1, 2)


// - Ejemplo 3:
void PrintFirstElement(string[] lista)
{
    Console.WriteLine(lista[0]);
}

PrintFirstElement(animales)


// - Ejemplo 4: Función quicksort:
List<int> quicksort(List<int> lista)
{
    if (lista.Count <= 1)
    {
        return lista;
    }

    int pivote = lista[0];
    List<int> izquierda = new List<int>();
    List<int> derecha = new List<int>();

    int i = 1;
    for (i; i < lista.Count; i++)
    {
        if (lista[i] < pivote) {
            izquierda.Add(lista[i]);
        }
        else
        {
            derecha.Add(lista[i]);
        }
    }

    List<int> primero = quicksort(izquierda);
    List<int> medio = new List<int> { pivote };
    List<int> segundo = quicksort(derecha);

    primero.AddRange(medio);
    primero.AddRange(segundo);
    return primero;
}

List<int> numeros2 = new List<int> { 23, 45, 16, 37, 3, 99,22 };
List<int> listaOrdenada = quicksort(numeros2);
Console.WriteLine(string.Join(", ", listaOrdenada));